### DISCUSSION TITLE
Ii it possible to make Toga objects with GTK objects, or the other way?

### DISCUSSION THREAD

**Discussion by Kuna42** [beeware/toga#2380] [Q&A]
I am thinking, it might be a good idea, if there is a translator for GTK objects into Toga objects.
The idea of designing the layout, see [this enchantment idea](https://github.com/beeware/toga/issues/2293#issuecomment-1868516951), might be easier if this is possible.

I ask for advice, if it is not possible, how i should implement it.

It is gtk specific, so the toga-gtk seems to be the right choice, but i want to do something like:

```python
# place this inside toga-core ?

# gtk for insert "*.glade" file
import gi
from gi.repository import Gtk

import toga.gtk_to_toga

def toga_layout_from_file(file: str):
    builder = Gtk.Builder()
    builder.add_from_file(file)
    handler = None
    builder.connect_signals(handler)
    return gtk_to_toga(builder)

```
I still have still no idea how this will look, but i think an important tool for something like this will be a class/object converter.

@freakboy3742 would you think something like a class converter is a good idea?

**Comment by freakboy3742** [comment]
I'm not sure exactly what you're proposing here. 

If you're proposing a mechanism for using `.glade` files (or any other existing markup format) as Toga's markup format, that's already been discussed on #2293. I can definitely see how such a parser might be useful, if only because it means Toga wouldn't need to develop it's own GUI designer tool. However, it's not going to be a simple mapping because the widgets, widget features and layout mechanisms of a non-Toga framework aren't necessarily going to map cleanly to Toga's API. It might be possible for simple cases, but for anything non-trivial you're going to hit friction in the conversion.

If you're proposing a mechanism for injecting a Glade layout into a Toga layout: I can see how that might work (and be relatively straightforward) - but it's also going to be a problem for Toga more generally, because Toga supports a lot of GUI frameworks that aren't GTK. At best, it would be a solution for the "I need to use this GTK widget in the GTK version of my Toga app".

That said - clearly documenting how to drop to the native layer and inject a block of raw GTK content into a toga-gtk app could be a useful contribution. However, I suspect this will be mostly a documentation issue (with a similar documentation need for every other backend), rather than something that requires specific helpers. There *might* be a need for a "NativeAdapter" widget (or something like that) that exposes a native container and simplifies the process of hosting native content - but I'm not necessarily convinced that would even be required.
